File: src/lib/designer/gradient.ts
import { Color } from "./color";

export class GradientPoint {
	// position on gradient
	t: number;

	// color of point
	color: Color;
}

export class Gradient {
	points: GradientPoint[];

	constructor() {
		this.points = [];
	}

	addPoint(t: number, color: Color): GradientPoint {
		const point = new GradientPoint();
		point.t = t;
		point.color = color;

		this.points.push(point);
		this.sort();

		return point;
	}

	removePoint(point: GradientPoint) {
		this.points.splice(this.points.indexOf(point), 1);
	}

	clear() {
		this.points = [];
	}

	sort() {
		this.points.sort(function(a: GradientPoint, b: GradientPoint) {
			return a.t - b.t;
		});
	}

	sample(t: number): Color {
		if (this.points.length == 0) return new Color();
		if (this.points.length == 1) return this.points[0].color.clone();

		// here at least two points are available
		if (t < this.points[0].t) return this.points[0].color.clone();

		const last = this.points.length - 1;
		if (t > this.points[last].t) return this.points[last].color.clone();

		// find two points and lerp
		for (let i = 0; i < this.points.length - 1; i++) {
			if (this.points[i + 1].t > t) {
				const p1 = this.points[i];
				const p2 = this.points[i + 1];

				const lerpPos = (t - p1.t) / (p2.t - p1.t);
				const color = new Color();
				color.copy(p1.color);
				color.lerp(p2.color, lerpPos);

				return color;
			}
		}

		// should never get to this point
		return new Color();
	}

	clone() {
		const grad = new Gradient();
		grad.clear();

		for (const p of this.points) {
			grad.addPoint(p.t, p.color.clone());
		}

		return grad;
	}

	public static parse(obj: any) {
		const gradient = new Gradient();
		for (const p of obj.points) {
			const t = p.t;
			const color = new Color();
			color.copy(p.color);

			gradient.addPoint(t, color);
		}

		return gradient;
	}

	public static default() {
		const gradient = new Gradient();
		gradient.addPoint(0, new Color(0, 0, 0, 1.0));
		gradient.addPoint(1, new Color(1, 1, 1, 1.0));

		return gradient;
	}
}

File: src/colorpicker/gradient.cpp
#include "gradient.h"

void Gradient::sort()
{
    std::sort(points.begin(), points.end(),
              [](const GradientPoint& a, const GradientPoint& b) {
                  return a.position < b.position;
              });
}

QColor Gradient::sample(float position) const
{
    // Default color is black
    QColor color = Qt::black;

    if (points.size() == 0) {
        return color;
    }

    if (points.size() == 1) {
        return points[0].color;
    }

    // Find surrounding points
    int leftIdx = -1, rightIdx = -1;
    for (int i = 0; i < points.size(); ++i) {
        if (points[i].position <= position) {
            leftIdx = i;
        }
        if (points[i].position >= position && rightIdx == -1) {
            rightIdx = i;
            break;
        }
    }

    if (leftIdx >= 0 && rightIdx >= 0 && leftIdx != rightIdx) {
        // Interpolate between the two colors
        float t = (position - points[leftIdx].position) /
                  (points[rightIdx].position - points[leftIdx].position);
        QColor leftColor = points[leftIdx].color;
        QColor rightColor = points[rightIdx].color;

        color = QColor(
            leftColor.red() + t * (rightColor.red() - leftColor.red()),
            leftColor.green() + t * (rightColor.green() - leftColor.green()),
            leftColor.blue() + t * (rightColor.blue() - leftColor.blue()));
    }
    else if (leftIdx >= 0) {
        color = points[leftIdx].color;
    }
    else if (rightIdx >= 0) {
        color = points[rightIdx].color;
    }

    return color;
}

Gradient Gradient::defaultGradient()
{
    Gradient gradient;
    gradient.addPoint(GradientPoint(0.0, QColor(0, 0, 0, 255)));
    gradient.addPoint(GradientPoint(1.0, QColor(255, 255, 255, 255)));
    return gradient;
}